// ngx-bootstrap/datepicker/esm5/themes/bs/bs-current-date-view.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, Input } from '@angular/core';
import * as ɵngcc0 from '@angular/core';

const _c0 = [1, "current-timedate"];
var BsCurrentDateViewComponent = /** @class */ (function () {
    function BsCurrentDateViewComponent() {
    }
    BsCurrentDateViewComponent.propDecorators = {
        title: [{ type: Input }]
    };
BsCurrentDateViewComponent.ngComponentDef = ɵngcc0.ɵɵdefineComponent({ type: BsCurrentDateViewComponent, selectors: [["bs-current-date"]], factory: function BsCurrentDateViewComponent_Factory(t) { return new (t || BsCurrentDateViewComponent)(); }, inputs: { title: "title" }, consts: 3, vars: 1, template: function BsCurrentDateViewComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵelementStart(0, "div", _c0);
        ɵngcc0.ɵɵelementStart(1, "span");
        ɵngcc0.ɵɵtext(2);
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
    } if (rf & 2) {
        ɵngcc0.ɵɵselect(2);
        ɵngcc0.ɵɵtextInterpolate(ctx.title);
    } }, encapsulation: 2 });
/*@__PURE__*/ ɵngcc0.ɵsetClassMetadata(BsCurrentDateViewComponent, [{
        type: Component,
        args: [{
                selector: 'bs-current-date',
                template: "<div class=\"current-timedate\"><span>{{ title }}</span></div>"
            }]
    }], function () { return []; }, { title: [{
            type: Input
        }] });
    return BsCurrentDateViewComponent;
}());
export { BsCurrentDateViewComponent };
if (false) {
    /** @type {?} */
    BsCurrentDateViewComponent.prototype.title;
}
