// ngx-bootstrap/collapse/esm5/collapse.module.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { NgModule } from '@angular/core';
import { CollapseDirective } from './collapse.directive';
import * as ɵngcc0 from '@angular/core';
var CollapseModule = /** @class */ (function () {
    function CollapseModule() {
    }
    /**
     * @return {?}
     */
    CollapseModule.forRoot = /**
     * @return {?}
     */
    function () {
        return { ngModule: CollapseModule, providers: [] };
    };
CollapseModule.ngModuleDef = ɵngcc0.ɵɵdefineNgModule({ type: CollapseModule });
/*@__PURE__*/ ɵngcc0.ɵɵsetNgModuleScope(CollapseModule, { declarations: function () { return [CollapseDirective]; }, exports: function () { return [CollapseDirective]; } });
/*@__PURE__*/ ɵngcc0.ɵsetClassMetadata(CollapseModule, [{
        type: NgModule,
        args: [{
                declarations: [CollapseDirective],
                exports: [CollapseDirective]
            }]
    }], function () { return []; }, null);
CollapseModule.ngInjectorDef = ɵngcc0.ɵɵdefineInjector({ factory: function CollapseModule_Factory(t) { return new (t || CollapseModule)(); } });
    return CollapseModule;
}());
export { CollapseModule };
